From 9797eb441b2dd6a8571cf43d535d24822c50c16e Mon Sep 17 00:00:00 2001
From: Baokun Li <libaokun1@huawei.com>
Date: Tue, 30 Nov 2021 16:32:05 +0800
Subject: [PATCH] ubifs: read-only if LEB may always be taken in
 ubifs_garbage_collect
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 9797eb441b2dd6a8571cf43d535d24822c50c16e


hulk inclusion
category: bugfix
bugzilla: 182993 https://gitee.com/openeuler/kernel/issues/I4DDEL

---------------------------

If ubifs_garbage_collect_leb() returns -EAGAIN and ubifs_return_leb
returns error, a LEB will always has a "taken" flag. In this case,
set the ubifs to read-only to prevent a worse situation.

Signed-off-by: Baokun Li <libaokun1@huawei.com>
Reviewed-by: Zhang Yi <yi.zhang@huawei.com>

Signed-off-by: Chen Jun <chenjun102@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/ubifs/gc.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/fs/ubifs/gc.c b/fs/ubifs/gc.c
index 238a60294d10..e43c8161423d 100644
--- a/fs/ubifs/gc.c
+++ b/fs/ubifs/gc.c
@@ -756,8 +756,14 @@ int ubifs_garbage_collect(struct ubifs_info *c, int anyway)
 				 * caller instead of the original '-EAGAIN'.
 				 */
 				err = ubifs_return_leb(c, lp.lnum);
-				if (err)
+				if (err) {
 					ret = err;
+					/* LEB may always be "taken". So set
+					 * the ubifs to read-only. Sync wbuf
+					 * will return -EROFS, then go "out".
+					 */
+					ubifs_ro_mode(c, ret);
+				}
 				/*  Maybe double return if go out */
 				lp.lnum = -1;
 				break;
-- 
2.26.2

